File: cloud/admin/src/routes/connections.js
// /api/connections — upstream provider credentials per tenant.
// Plaintext credentials are encrypted with the per-tenant DEK before storage
// and never read back (the response only describes shape).

import { query, encryptForTenant, NotFoundError, ValidationError, getRedis } from '@9router-cloud/shared';
import { readJson, ok, noContent } from '../lib/http.js';
import { requireSession, requireRole } from '../middleware/sessionAuth.js';

export async function listConnections(req, res) {
  const s = requireSession(req);
  const { rows } = await query(`
    SELECT id, provider, name, auth_type, metadata, enabled, weight, oauth_expires_at, created_at, updated_at
    FROM connections
    WHERE tenant_id = $1
    ORDER BY provider ASC, id ASC
  `, [s.tenantId]);
  ok(res, { items: rows.map(serialize) });
}

export async function createConnection(req, res) {
  const s = requireSession(req);
  requireRole(s, 'owner', 'admin');
  const body = await readJson(req);

  const provider = required(body, 'provider').toLowerCase().trim();
  const name = required(body, 'name').trim();
  const authType = (body.authType || 'api_key').toLowerCase();
  if (!['api_key', 'oauth', 'passthrough'].includes(authType)) {
    throw new ValidationError(`authType must be one of api_key|oauth|passthrough`);
  }
  if (!body.credentials || typeof body.credentials !== 'object') {
    throw new ValidationError('credentials object is required');
  }
  if (authType === 'api_key' && !body.credentials.api_key) {
    throw new ValidationError('credentials.api_key is required for authType=api_key');
  }
  const metadata = body.metadata && typeof body.metadata === 'object' ? body.metadata : {};
  const weight = Number.isInteger(body.weight) ? body.weight : 1;
  const enabled = body.enabled !== false;

  const blob = encryptForTenant(s.tenantId, body.credentials);
  let row;
  try {
    const r = await query(`
      INSERT INTO connections (tenant_id, provider, name, auth_type, credentials_encrypted, metadata, enabled, weight)
      VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
      RETURNING id, provider, name, auth_type, metadata, enabled, weight, oauth_expires_at, created_at, updated_at
    `, [s.tenantId, provider, name, authType, blob, metadata, enabled, weight]);
    row = r.rows[0];
  } catch (e) {
    if (e.code === '23505') {
      throw new ValidationError(`Connection (${provider}, ${name}) already exists`);
    }
    throw e;
  }
  await invalidateConnCache(s.tenantId, provider);
  ok(res, serialize(row), 201);
}

export async function updateConnection(req, res, { params }) {
  const s = requireSession(req);
  requireRole(s, 'owner', 'admin');
  const id = Number(params.id);
  const body = await readJson(req);

  // Fetch current (for tenant scoping + provider for cache invalidation)
  const cur = await query(`SELECT id, provider FROM connections WHERE id = $1 AND tenant_id = $2`, [id, s.tenantId]);
  if (cur.rowCount === 0) throw new NotFoundError('Connection not found');
  const provider = cur.rows[0].provider;

  const sets = [];
  const args = [];
  let i = 1;
  if (body.name !== undefined) { sets.push(`name = $${i++}`); args.push(String(body.name).trim()); }
  if (body.metadata !== undefined) { sets.push(`metadata = $${i++}`); args.push(body.metadata); }
  if (body.enabled !== undefined) { sets.push(`enabled = $${i++}`); args.push(!!body.enabled); }
  if (Number.isInteger(body.weight)) { sets.push(`weight = $${i++}`); args.push(body.weight); }
  if (body.credentials && typeof body.credentials === 'object') {
    sets.push(`credentials_encrypted = $${i++}`);
    args.push(encryptForTenant(s.tenantId, body.credentials));
  }
  if (sets.length === 0) throw new ValidationError('No updatable fields supplied');
  sets.push(`updated_at = now()`);
  args.push(id, s.tenantId);
  const { rows } = await query(
    `UPDATE connections SET ${sets.join(', ')}
     WHERE id = $${i++} AND tenant_id = $${i}
     RETURNING id, provider, name, auth_type, metadata, enabled, weight, oauth_expires_at, created_at, updated_at`,
    args
  );
  await invalidateConnCache(s.tenantId, provider);
  ok(res, serialize(rows[0]));
}

export async function deleteConnection(req, res, { params }) {
  const s = requireSession(req);
  requireRole(s, 'owner', 'admin');
  const id = Number(params.id);
  const { rows } = await query(
    `DELETE FROM connections WHERE id = $1 AND tenant_id = $2 RETURNING provider`,
    [id, s.tenantId]
  );
  if (rows.length === 0) throw new NotFoundError('Connection not found');
  await invalidateConnCache(s.tenantId, rows[0].provider);
  noContent(res);
}

async function invalidateConnCache(tenantId, provider) {
  await getRedis().del(`conn_cache:${tenantId}:${provider}`);
}

function required(obj, field) {
  if (obj[field] === undefined || obj[field] === null || obj[field] === '') {
    throw new ValidationError(`Missing field: ${field}`);
  }
  return String(obj[field]);
}

function serialize(row) {
  return {
    id: Number(row.id),
    provider: row.provider,
    name: row.name,
    authType: row.auth_type,
    metadata: row.metadata || {},
    enabled: row.enabled,
    weight: row.weight,
    oauthExpiresAt: row.oauth_expires_at,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

File: internal/admin/connections.go
package admin

import (
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"github.com/amaozhao/lazirouter/internal/cryptox"
	"github.com/amaozhao/lazirouter/internal/db"
	"github.com/amaozhao/lazirouter/internal/errs"
	"github.com/amaozhao/lazirouter/internal/httpx"
	"github.com/amaozhao/lazirouter/internal/picker"
	"github.com/jackc/pgx/v5"
)

type connDTO struct {
	ID              int64           `json:"id"`
	Provider        string          `json:"provider"`
	Name            string          `json:"name"`
	AuthType        string          `json:"authType"`
	Metadata        json.RawMessage `json:"metadata"`
	Enabled         bool            `json:"enabled"`
	Weight          int             `json:"weight"`
	OauthExpiresAt  *time.Time      `json:"oauthExpiresAt"`
	CreatedAt       time.Time       `json:"createdAt"`
	UpdatedAt       time.Time       `json:"updatedAt"`
}

func (d *Deps) ListConnections(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	rows, err := db.Pool().Query(r.Context(), `
		SELECT id, provider, name, auth_type, metadata, enabled, weight, oauth_expires_at, created_at, updated_at
		FROM connections WHERE tenant_id = $1 ORDER BY id
	`, s.TenantID)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	defer rows.Close()
	items := []connDTO{}
	for rows.Next() {
		var c connDTO
		if err := rows.Scan(&c.ID, &c.Provider, &c.Name, &c.AuthType, &c.Metadata, &c.Enabled, &c.Weight, &c.OauthExpiresAt, &c.CreatedAt, &c.UpdatedAt); err != nil {
			httpx.WriteError(w, err)
			return
		}
		items = append(items, c)
	}
	httpx.WriteJSON(w, http.StatusOK, map[string]any{"items": items})
}

type createConnReq struct {
	Provider    string         `json:"provider"`
	Name        string         `json:"name"`
	AuthType    string         `json:"authType"`
	Credentials map[string]any `json:"credentials"`
	Metadata    map[string]any `json:"metadata"`
	Weight      *int           `json:"weight"`
}

func (d *Deps) CreateConnection(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	var req createConnReq
	if err := httpx.ReadJSON(r, &req); err != nil {
		httpx.WriteError(w, err)
		return
	}
	if req.Provider == "" || req.Name == "" || req.AuthType == "" {
		httpx.WriteError(w, errs.Validation("provider, name, authType required"))
		return
	}
	if req.Credentials == nil {
		httpx.WriteError(w, errs.Validation("credentials object is required"))
		return
	}
	if req.AuthType == "api_key" {
		if _, ok := req.Credentials["api_key"]; !ok {
			httpx.WriteError(w, errs.Validation("credentials.api_key is required for authType=api_key"))
			return
		}
	}
	weight := 1
	if req.Weight != nil && *req.Weight >= 1 {
		weight = *req.Weight
	}
	enc, err := cryptox.EncryptForTenant(d.Cfg.MasterKey, s.TenantID, req.Credentials)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	metaJSON := []byte("{}")
	if req.Metadata != nil {
		metaJSON = marshalJSON(req.Metadata)
	}
	var c connDTO
	err = db.Pool().QueryRow(r.Context(), `
		INSERT INTO connections (tenant_id, provider, name, auth_type, credentials_encrypted, metadata, weight)
		VALUES ($1, $2, $3, $4, $5, $6, $7)
		RETURNING id, provider, name, auth_type, metadata, enabled, weight, oauth_expires_at, created_at, updated_at
	`, s.TenantID, req.Provider, req.Name, req.AuthType, enc, metaJSON, weight).Scan(
		&c.ID, &c.Provider, &c.Name, &c.AuthType, &c.Metadata, &c.Enabled, &c.Weight,
		&c.OauthExpiresAt, &c.CreatedAt, &c.UpdatedAt,
	)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	_ = picker.InvalidateCache(r.Context(), s.TenantID, req.Provider)
	httpx.WriteJSON(w, http.StatusOK, c)
}

type patchConnReq struct {
	Name        *string                 `json:"name"`
	Weight      *int                    `json:"weight"`
	Enabled     *bool                   `json:"enabled"`
	Metadata    *map[string]any         `json:"metadata"`
	Credentials *map[string]any         `json:"credentials"`
}

func (d *Deps) UpdateConnection(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	id, err := pathInt(r, "id")
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	var req patchConnReq
	if err := httpx.ReadJSON(r, &req); err != nil {
		httpx.WriteError(w, err)
		return
	}
	// Pre-check ownership + grab provider for cache invalidation
	var provider string
	err = db.Pool().QueryRow(r.Context(),
		`SELECT provider FROM connections WHERE id = $1 AND tenant_id = $2`, id, s.TenantID).Scan(&provider)
	if errors.Is(err, pgx.ErrNoRows) {
		httpx.WriteError(w, errs.NotFound("Connection not found"))
		return
	}
	if err != nil {
		httpx.WriteError(w, err)
		return
	}

	// Build dynamic UPDATE
	sets := []string{}
	args := []any{}
	add := func(col string, val any) {
		args = append(args, val)
		sets = append(sets, col+" = $"+itoa(len(args)))
	}
	if req.Name != nil {
		add("name", *req.Name)
	}
	if req.Weight != nil {
		add("weight", *req.Weight)
	}
	if req.Enabled != nil {
		add("enabled", *req.Enabled)
	}
	if req.Metadata != nil {
		add("metadata", marshalJSON(*req.Metadata))
	}
	if req.Credentials != nil {
		enc, err := cryptox.EncryptForTenant(d.Cfg.MasterKey, s.TenantID, *req.Credentials)
		if err != nil {
			httpx.WriteError(w, err)
			return
		}
		add("credentials_encrypted", enc)
	}
	if len(sets) == 0 {
		httpx.WriteError(w, errs.Validation("no updatable fields"))
		return
	}
	args = append(args, id, s.TenantID)
	q := `UPDATE connections SET ` + join(sets, ", ") + `, updated_at = now()
	      WHERE id = $` + itoa(len(args)-1) + ` AND tenant_id = $` + itoa(len(args)) + `
	      RETURNING id, provider, name, auth_type, metadata, enabled, weight, oauth_expires_at, created_at, updated_at`
	var c connDTO
	err = db.Pool().QueryRow(r.Context(), q, args...).Scan(
		&c.ID, &c.Provider, &c.Name, &c.AuthType, &c.Metadata, &c.Enabled, &c.Weight,
		&c.OauthExpiresAt, &c.CreatedAt, &c.UpdatedAt,
	)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	_ = picker.InvalidateCache(r.Context(), s.TenantID, provider)
	httpx.WriteJSON(w, http.StatusOK, c)
}

func (d *Deps) DeleteConnection(w http.ResponseWriter, r *http.Request) {
	s, err := RequireSession(r, d.Cfg)
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	id, err := pathInt(r, "id")
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	var provider string
	err = db.Pool().QueryRow(r.Context(),
		`DELETE FROM connections WHERE id = $1 AND tenant_id = $2 RETURNING provider`, id, s.TenantID).Scan(&provider)
	if errors.Is(err, pgx.ErrNoRows) {
		httpx.WriteError(w, errs.NotFound("Connection not found"))
		return
	}
	if err != nil {
		httpx.WriteError(w, err)
		return
	}
	_ = picker.InvalidateCache(r.Context(), s.TenantID, provider)
	httpx.WriteJSON(w, http.StatusOK, map[string]any{"ok": true, "id": id})
}

// tiny local helpers — avoid adding strconv/strings importance noise to every call site
func itoa(n int) string {
	if n < 10 {
		return string(rune('0' + n))
	}
	return slowItoa(n)
}
func slowItoa(n int) string {
	if n == 0 {
		return "0"
	}
	s := ""
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		s = string(rune('0'+(n%10))) + s
		n /= 10
	}
	if neg {
		s = "-" + s
	}
	return s
}
func join(parts []string, sep string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += sep
		}
		out += p
	}
	return out
}
